fusequery/query/src/optimizers/optimizer_constant_folding.rs

```diff
@@ -2,132 +2,191 @@
-            },
-            _ => Expression::BinaryExpression { left, op, right },
-        },
-        expr => {
-            // do nothing
-            expr
+impl ConstantFoldingImpl {
+    fn rewrite_alias(alias: &str, expr: Expression) -> Result<Expression> {
+        Ok(Expression::Alias(alias.to_string(), Box::new(expr)))
+    }
+
+    fn constants_arguments(args: &[Expression]) -> bool {
+        !args
+            .iter()
+            .any(|expr| !matches!(expr, Expression::Literal { .. }))
+    }
+
+    fn rewrite_function<F>(op: &str, args: Expressions, name: String, f: F) -> Result<Expression>
+    where F: Fn(&str, Expressions) -> Expression {
+        let function = FunctionFactory::get(op)?;
+
+        if function.is_deterministic() && ConstantFoldingImpl::constants_arguments(&args) {
+            let op = op.to_string();
+            return ConstantFoldingImpl::execute_expression(
+                Expression::ScalarFunction { op, args },
+                name,
+            );
         }
-    };
-    Ok(new_expr)
+
+        Ok(f(op, args))
+    }
+
+    fn create_scalar_function(op: &str, args: Expressions) -> Expression {
+        let op = op.to_string();
+        Expression::ScalarFunction { op, args }
+    }
+
+    fn create_unary_expression(op: &str, mut args: Expressions) -> Expression {
+        let op = op.to_string();
+        let expr = Box::new(args.remove(0));
+        Expression::UnaryExpression { op, expr }
+    }
+
+    fn create_binary_expression(op: &str, mut args: Expressions) -> Expression {
+        let op = op.to_string();
+        let left = Box::new(args.remove(0));
+        let right = Box::new(args.remove(0));
+        Expression::BinaryExpression { op, left, right }
+    }
+
+    fn expr_executor(schema: &DataSchemaRef, expr: Expression) -> Result<ExpressionExecutor> {
+        let output_fields = vec![expr.to_data_field(schema)?];
+        let output_schema = DataSchemaRefExt::create(output_fields);
+        ExpressionExecutor::try_create(
+            "Constant folding optimizer.",
+            schema.clone(),
+            output_schema,
+            vec![expr],
+            false,
+        )
+    }
+
+    fn execute_expression(expression: Expression, origin_name: String) -> Result<Expression> {
+        let input_fields = vec![DataField::new("_dummy", DataType::UInt8, false)];
+        let input_schema = Arc::new(DataSchema::new(input_fields));
+
+        let expression_executor = Self::expr_executor(&input_schema, expression)?;
+        let dummy_columns = vec![DataColumn::Constant(DataValue::UInt8(Some(1)), 1)];
+        let data_block = DataBlock::create(input_schema, dummy_columns);
+        let executed_data_block = expression_executor.execute(&data_block)?;
+
+        ConstantFoldingImpl::convert_to_expression(origin_name, executed_data_block)
+    }
+
+    fn convert_to_expression(column_name: String, data_block: DataBlock) -> Result<Expression> {
+        assert_eq!(data_block.num_rows(), 1);
```

Comment:
Is it possible to handle errors better?

Reply:
It can only return one row result, I think we should be more strict